cloudify_azure/resources/deployment.py

```diff
@@ -163,3 +148,130 @@ def delete(ctx, **_):
+    resource_group_name, deployment_name, api_version = \
+        get_resource_group_name_deployment_name_and_api_version(ctx)
+
+    resource_group = ResourceGroup(azure_config, ctx.logger, api_version)
+    try:
+        resource_group.get(resource_group_name)
+    except CloudError:
+        ctx.logger.info("Resource group {rg_name} does not exist. State will "
+                        "be empty.".format(rg_name=resource_group_name))
+        ctx.instance.runtime_properties[STATE] = []
+        ctx.instance.runtime_properties[IS_DRIFTED] = True
+        return
+
+    # Get the resources list that crated during the template run.
+    initial_resources = ctx.instance.runtime_properties.get(
+        'resource', {}).get('properties', {}).get('output_resources', [])
+    ctx.logger.debug("initial_resources: {}".format(initial_resources))
+    actual_resources = resource_group.list_resources(resource_group_name)
+
+    deployment = Deployment(azure_config, ctx.logger)
+    properties, params = get_properties_and_formated_params(ctx, **kwargs)
+    template = ctx.instance.runtime_properties.get('template') or get_template(
+        ctx, properties)
+    # some resources are nested like subnets so they do not appear in
+    # actual_resources. We will search them using the what-if result.
+
+    what_if_res = execute_what_if(deployment,
+                                  resource_group_name,
+                                  deployment_name,
+                                  template,
+                                  params)
+    calculate_state(ctx, initial_resources, actual_resources, what_if_res)
+
+
+def get_properties_and_formated_params(ctx, **kwargs):
+    properties = {}
+    properties.update(ctx.node.properties)
+    properties.update(kwargs)
+    params = format_params(properties.get('params', {}))
+    return properties, params
+
+
+def get_resource_group_name_deployment_name_and_api_version(ctx):
+    deployment_name = utils.get_resource_name(ctx)
+    resource_group_name = ctx.node.properties.get(
+        'resource_group_name', deployment_name)
+    api_version = \
+        ctx.node.properties.get('api_version', constants.API_VER_RESOURCES)
+    return deployment_name, resource_group_name, api_version
+
+
+def execute_what_if(deployment,
+                    resource_group_name,
+                    deployment_name,
+                    template,
+                    params):
+    what_if_properties = {
+        'mode': DeploymentMode.incremental,
+        'template': template,
+        'parameters': params
+    }
+    try:
+        return deployment.what_if(resource_group_name,
+                                  deployment_name,
+                                  what_if_properties)
+    except CloudError:
+        raise cfy_exc.NonRecoverableError(
+            "What if operation on deployment {dep_name} failed. Can't "
+            "calculate accurate state.".format(dep_name=deployment_name))
```

Comment:
Are there any runtime_properties that need to be set on a fatal error like this?

Reply:
No, because if we cant pull the state so we will leave it as is.